src/main/java/ch/uzh/ifi/hase/soprafs22/screwyourneighborserver/controller/FollowUpGameController.java

```diff
@@ -0,0 +1,69 @@
+package ch.uzh.ifi.hase.soprafs22.screwyourneighborserver.controller;
+
+import ch.uzh.ifi.hase.soprafs22.screwyourneighborserver.entity.Game;
+import ch.uzh.ifi.hase.soprafs22.screwyourneighborserver.entity.GameState;
+import ch.uzh.ifi.hase.soprafs22.screwyourneighborserver.repository.GameRepository;
+import ch.uzh.ifi.hase.soprafs22.screwyourneighborserver.sideeffects.GameEventHandler;
+import javax.transaction.Transactional;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.rest.webmvc.support.RepositoryEntityLinks;
+import org.springframework.hateoas.EntityModel;
+import org.springframework.hateoas.Link;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.client.HttpClientErrorException;
+
+@RestController
+@Transactional
+public class FollowUpGameController {
+  private final GameRepository gameRepository;
+  private final GameEventHandler gameEventHandler;
+
+  private final RepositoryEntityLinks repositoryEntityLinks;
+
+  @Autowired
+  public FollowUpGameController(
+      GameRepository gameRepository,
+      GameEventHandler gameEventHandler,
+      RepositoryEntityLinks repositoryEntityLinks) {
+    this.gameRepository = gameRepository;
+    this.gameEventHandler = gameEventHandler;
+    this.repositoryEntityLinks = repositoryEntityLinks;
+  }
+
+  @PostMapping("/games/{gameId}/nextGame")
+  public ResponseEntity<EntityModel<Game>> createNextGame(@PathVariable Long gameId) {
+    Game previousGame =
+        gameRepository
+            .findById(gameId)
+            .orElseThrow(
+                () ->
+                    new HttpClientErrorException(
+                        HttpStatus.NOT_FOUND,
+                        "The game with the id %s does not exist.".formatted(gameId)));
+
+    if (previousGame.getNextGame() != null) {
+      throw new HttpClientErrorException(
+          HttpStatus.UNPROCESSABLE_ENTITY, "nextGame already exists");
```

Comment:
Du denkst schon an alles ;)
Aber: wenn die gleichzeitig auf den Button "next Game" klicken, dann können zwei Games erstellt werden oder? Siehe Kommentar bei Zeile 58-61.

Reply:
Sollte nicht, ist @Transactional: der erste kommt durch, der zweite nicht.

Reply:
Ich verstehe das anders. @transactional sorgt doch nur dafür, dass alle diese DB-Operationen als eine atomare Einheit (Transaktion) ausgeführt werden.
Wenn nun eine zweite Transaktion T2 nextGame gestartet wird von einem anderen Player, dann wird bei der Prüfung das Feld nextGame (noch) leer sein, wenn die erste Transaktion T1 nicht beendet ist, somit wird T2 weitermachen, durchlaufen, zum Schluss nextGame verlinken und speichern (und dabei den von der inzwischen beendeten Transaktion T1 gesetzten Wert überschreiben).

Reply:
Es gäbe eine einfache Lösung: man könnte vom game eine Rückreferenz auf das letzte Spiel, z.B. lastGame oder einfach lastGameID machen, die mit einem unique constraint belegen und dann stellt die DB sicher, dass nie zwei Nachfolgespiele für ein Spiel erstellt werden.

Reply:
Es ist schon lange her, aber soweit ich mich erinnern kann erstellt der read innerhalb einer Transaktion einen read lock auf der row. Wenn jetzt jemand anders während deiner Transaktion die row ändern möchte, bekommt er keinen write lock, weil schon ein read lock darauf ist. Was je nach dem sein könnte ist dass das zweite update einfach verzögert wird, und dann trotzdem überschreibt.
Man könnte noch sicherer gehen und bei der Query am Anfang den intend "FOR UPDATE" mitgeben (ohne ORM halt), damit der zweite das sicher nicht überschreiben kann. Ich denke aber das ist gut genug für unseren use case.

(https://www.baeldung.com/jpa-pessimistic-locking)

Reply:
Ja genau, ich hätte entweder das von dir beschriebene Szenario mit einem dirty read und anschliessendem verzögerten write erwartet oder dass die T1 nicht mal einen read-lock hielte und T2 somit einfach dirty lesen und dann auch gleich schreiben könnte. Aber in beiden Szenarien wird überschrieben.

Ich habe verstanden, dass man in JPA setLockMode(LockModeType.PESSIMISTIC_WRITE) setzen muss um das Äquivalent von "FOR UPDATE" zu erreichen? (https://mobiarch.wordpress.com/2013/08/01/doing-select-for-update-in-jpa/)

Reply:
Haben wir mit implicit optimistic locking gemacht mit einem @Version attribute auf game